### ISSUE TITLE
aws_secretsmanager_secret does not rebuild deleted secret

### ISSUE THREAD

**Issue by zarlant** [terraform-providers/terraform-provider-aws#5127]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
Terraform v0.11.7
+ provider.aws v1.25.0

### Affected Resource(s)
* aws_secretsmanager_secret

### Terraform Configuration Files

```hcl
resource "aws_secretsmanager_secret" "testsecret" {
  name = "mycool_secret"
}
```

### Expected Behavior
Terraform should look to see if the secret exists and create it.

### Actual Behavior
Terraform sees the secret id in the state file and does not verify that the secret still exists in AWS

### Steps to Reproduce
1. `terraform apply`
2. Manually delete secret using the CLI, or some other means
3. `terraform plan` or `terraform apply`
4. Terraform thinks there is nothing to do

### Important Factoids
This only seems to happen if the secret is deleted outside of terraform.

**Comment by bflad** [comment]
Relevant Secrets Manager documentation: https://docs.aws.amazon.com/secretsmanager/latest/userguide/manage_delete-restore-secret.html

Deleting a secret only marks it as deprecated and does not immediately fully delete it. We should probably add a documentation note about this in the [resource documentation](https://www.terraform.io/docs/providers/aws/r/secretsmanager_secret.html).

We might be able to mark these deprecated secrets for recreation in the resource read function, but the creation in that scenario I believe will always fail if the secret name is the same. We'll need to check the user experience of that (e.g. it returns a helpful error when a deprecated secret already exists) or if we should return our own error about deprecated secrets instead.

A workaround is to create a different secret in the deprecation period.

**Comment by jangeja94** [comment]
If terraform checks to see if the key is scheduled for deletion before trying to create, this error would be resolved. This would be very helpful especially for people first starting out with terraform.

**Comment by hleb-rubanau** [comment]
>> This only seems to happen if the secret is deleted outside of terraform.

I believe it also happens in case of "terraform destroy" followed by "terraform apply".

Maybe appropriate action would be to use "undelete" API call instead of "create" if secret is scheduled for deletion.

This logic may not necessary be default. However, having a flag on terraform secret resource (enabling automatical "undelete" behaviour) would be very nice thing to have.

Otherwise, terraform runs are not 100% idempotent. (I.e. I cannot wipe my testing infrastructure in the end of the day to just recreate it next morning).

**Comment by hleb-rubanau** [comment]
Uh, got it. My suggestion won't work out.

Terraform has no idea of secret existence because after deletion secret's ARN is no more in the terraform state. So, prior to any planning secret should be **imported** into the state.

Any ideas to automate this process should not be considered neither suggested -- as implicit import will break the fundamental concepts of Terraform (it may lead to the situation where Alice and Bob compete over the control of same resource just because of using same name).

I solved the problem with little manual overhead, fairly acceptable:
1) Manually "cancelled deletion" for affected secrets  in the AWS management console
2) Performed "terraform import" for affected resources

Maybe step 1 was not even necessary, but I do not have time to check it now.

**Comment by bflad** [comment]
Pull request submitted with support of the new `ForceDeleteWithoutRecovery` parameter for `DeleteSecret`: #5583

**Comment by bflad** [comment]
This has been released in [version 1.33.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1330-august-22-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by neurogenesis** [comment]
@bflad : Resource policies aren't replaced when updated from a rendered template, using `terraform-provider-aws_v2.29.0_x4` (macos). The plan shows that it wants to update them, but the apply only updates the template data in the statefile (and not the resource policy itself).

In other words, a `terraform apply` identifies that the template changed, but the `terraform apply` only modifies the state file (and not the actual resource in AWS).

Policy applies seems to work on new secrets, but not on deleted / restored secrets. One can get into a weird state if a change replaces the secret and you have to restore it and subsequently modify it.

I thought about opening a new issue, but the issue name seems to reflect this accurately.
